_posts/2020-05-28-ui-5.md

```diff
@@ -0,0 +1,298 @@
+
+## Layering components
+
+We've changed the set of components used to produce various types of overlays. This is because the underlying logic has been improved and to clarify the scope of certain components that did a bit too much. We now have: `Layer`, `ComponentCover` and `CenterContent`. They replace the `Backdrop` and the `ScreenCover`, which have been removed.
+
+### Layer
+
+`Layer` is an overlay component that fills the entire viewport and allows you to stack various components on top of one another. The `Layer` accepts an `onClick` callback, so you can catch clicks on the background of whatever you're rendering. It also has a `translucent` prop, if you don't want the `Layer` to darken the background. An example:
+
+```jsx
+<Layer>
+    <p>This will be rendered on top of the app, regardless of where you place it in your markup</p>
+</Layer>
+<p>Text behind the layer</p>
+```
+
+The `Layer` uses React context internally to control the stacking logic. This context has been exposed via the `useLayerContext` hook, which can be used to append portals to the current layer-node, for advanced users.
+
+### ComponentCover
+
+The `ComponentCover` is similar to the `Layer`, except `ComponentCover` only fills its parent, provided that the parent has a `relative` or `absolute` position. The `ComponentCover` also accepts an `onClick` and `translucent` prop, just like the `Layer`. An example of the `ComponentCover`:
+
+```jsx
+<div style={{ position: 'relative' }}>
+    <ComponentCover>
+        <p>
+            The ComponentCover will fill the parent div, and render this
+            paragraph on top of the div visually.
+        </p>
+    </ComponentCover>
+</div>
+```
+
+### CenterContent
+
+CenterContent is a component that centers its children. It has a `position` prop which can be used to vertically align the children at the `top`, `middle` (default), or `bottom`. It can be useful when you want to render a loading spinner on top of your app for example:
+
+```jsx
+// This will render a spinner on top of your app, centered in the middle:
+<Layer>
+    <CenteredContent>
+        <CircularLoader />
+    </CenteredContent>
+</Layer>
+```
+
+## Click based Menu
+
+The `Menu` has been refactored to make it easier to use. It is now click-based instead of hover-based, so sub-menus will stay open even if the user is no longer hovering over them.
+
+Besides that, the `MenuList` has been renamed to `Menu`, this component will expand to fill the full width of the parent container. It is a good fit for a sidebar menu for example. The original `Menu` component has been renamed to `FlyoutMenu`, to clearly indicate that this will render a flyout menu. An example of both:
+
+```jsx
+// So this will render full-width
+<Menu>
+    <MenuItem label="Item 1" />
+    <MenuItem label="Item 2" />
+</Menu>
+
+// And this will render as a popup
+<FlyoutMenu>
+    <MenuItem label="Item 1" />
+    <MenuItem label="Item 2" />
+</FlyoutMenu>
+```
```

Comment:
The reality is slightly more complicated than this. I think the text is fine, but the example could be tweaked a bit. Two points of importance:
- This comment not fully correct: `// And this will render as a popup`. In fact,  the "root" flyout-menu is not is a popoup, it is in a `Card`. Nested menus / submenus will fly-out in new Cards. 
- Since the current example is only one level deep, it doesn't really showcase the fly-out sub-menus.

Perhaps we could tweak like this:
```jsx
// This will render in a Card with width/height restrictions
<FlyoutMenu>
    <MenuItem label="Item 1">
        {/* This "sub-menu" will be rendered in a card in a popover */}
        <MenuItem label="Item 1a" />
        <MenuItem label="Item 1a" />
    </MenuItem>
    <MenuItem label="Item 2" />
</FlyoutMenu>
```

Reply:
Oh, I hadn't noticed the section that follows and explains how to make sub-menus.... Upon reflection, perhaps just updating the comment above would be better...
So just change `// And this will render as a popup` to `// And this will render in a Card with width/height restrictions`

Reply:
What do you think of this: 512c686765c14890989f3f9a41192187b3c3441e